homeassistant/components/weather/ambient.py

```diff
@@ -0,0 +1,291 @@
+
+    @property
+    def pressure(self):
+        """Return the pressure."""
+        if self._ambient.adoutput is not None:
+            return self._ambient.adoutput.pressure
+        return None
+
+    @property
+    def condition(self):
+        """Return the weather condition."""
+        if self._ambient.adoutput.hourly_rain > 0.1:
+            return 'pouring'
+
+        if self._ambient.adoutput.hourly_rain > 0:
+            return 'rainy'
+
+        if self._ambient.adoutput.daily_rain > 0:
+            return 'cloudy'
+
+        if self._ambient.adoutput.wind_gust > 15:
+            return 'windy'
+
+        if self._ambient.adoutput.solar_radiation >= 100:
+            return 'sunny'
+
+        return None
+
+    @property
+    def ozone(self):
+        """Return the ozone level."""
+        return self._ambient.adoutput.ozone
+
+    @property
+    def attribution(self):
+        """Return the attribution."""
+        attr_str = 'Data from Ambient Weather'
+        if self._ambient.station_name is not None:
+            attr_str += ' Station ' + self._ambient.station_name
+        if self._ambient.station_location is not None:
+            attr_str += '@' + self._ambient.station_location
+        return attr_str
+
+    @property
+    def forecast(self):
+        """Return the forecast. Ambient doesn't support this."""
+        return None
+
+    @property
+    def state_attributes(self):
+        """Return the state attributes."""
+        if self._ambient.adoutput is None:
+            return None
+
+        data = {}
+        for key in range(len(KEY_ORDER)):
+            data[KEY_ORDER[key]] = self._ambient.adoutput[key]
+
+        return data
+
+    def update(self):
+        """Update the data by delegating to the data object."""
+        self._ambient.update()
+
+
+class AmbientData:
+    """Helper object to retrieve Ambient data."""
+
+    def __init__(self, hass, api_key):
+        """Set up the http session and data structures."""
+        self.hass = hass
+        self._session = Session()
+        self._session.params = {
+            'applicationKey': APPLICATION_KEY,
+            'apiKey': api_key
+        }
+
+        self._data = namedtuple(
+            'status', KEY_ORDER
+            )
+        self.adoutput = None
+        self.station_name = None
+        self.station_location = None
+
+    def mph_to_kph(self, value_mph):
+        """Convert mph to metric if required."""
+        conv_mph = value_mph
+        if self.hass.config.units.is_metric:
+            conv_mph = round(conv_mph * MPH_TO_KPH, 1)
+        return conv_mph
+
+    def in_to_cm(self, value_in):
+        """Convert inches to metric if required."""
+        conv_in = value_in
+        if self.hass.config.units.is_metric:
+            conv_in = round(conv_in * IN_TO_CM, 2)
+        return conv_in
+
+    def update(self):
+        """Retrieve the data from Ambient and decode."""
+        try:
+            resp = self._session.get(AMBIENT_URL)
```

Comment:
The device interface, I/O and data parsing, needs to be extracted to standalone library published on pypi.

https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#6-communication-with-devices-services

Reply:
I updated and pushed my changes back but now I have all of the other deltas in my pull request. I clearly did something wrong. Sorry, git continues to be largely incomprehensible to me.